WIP: Nodegroup Scaling

richardcase opened the pull request:
### Description
This is currently work in progress and is dependent on a [change to goformation](https://github.com/awslabs/goformation/pull/103) so that ASG can be correctly unmarshalled.

It builds succesfully but it does not work yet.

Issue #116 

### Checklist
- [x] Code compiles correctly (i.e `make build`)
- [ ] Added tests that cover your change (if possible)
- [ ] All tests passing (i.e. `make test`)
- [ ] Added/modified documentation as required (such as the README)
- [ ] Added yourself to the `humans.txt` file

richardcase commented:
This is currently failing because of unmarshalling issues. I think #210 may help here and unblock this.

errordeveloper commented:
> This is currently failing because of unmarshalling issues.

Would you mind to expand please?

>  I think #210 may help here and unblock this.

Yes, potentially, but it'd be good to understand what is the issue with current version.

richardcase commented:
> Would you mind to expand please?

Sure. I'll have to re-acquaint myself with the issue i was seeing. Will get back to you asap.

richardcase commented:
@errordeveloper - it appears that when i [parse the template JSON](https://github.com/weaveworks/eksctl/blob/scale-116/pkg/cfn/manager/nodegroup.go#L55) thats returned from AWS that the properties aren't being populated:

<img width="1006" alt="screen shot 2018-09-26 at 13 32 40" src="https://user-images.githubusercontent.com/198425/46080188-23cf3180-c191-11e8-9037-7b55eb90feed.png">

I'm probably doing something wrong here.

richardcase commented:
@errordeveloper - it would be good to get your views on the following:

**Goformation**
I removed the use of Goformation for the time being (issue with Refs) and am manipulating the template with maps:
https://github.com/weaveworks/eksctl/blob/9459a0c503f98434fb691b4cba64eae361bcd20e/pkg/cfn/manager/nodegroup.go#L64

**Changesets**
I changed the code to use ChangeSets instead of updating the stack directly:
https://github.com/weaveworks/eksctl/blob/9459a0c503f98434fb691b4cba64eae361bcd20e/pkg/cfn/manager/api.go#L82

Also, the code is very much a WIP and needs a lot of refactoring (but it does scale). I'm going to start refactoringthe code now.

richardcase commented:
I'm also going to add waiting for the status of the changeset, listing the things that will change and then when executing the changeset we will wait for the required status.

errordeveloper commented on pkg/cfn/manager/nodegroup.go:
```diff
@@ -28,6 +34,50 @@ func (c *StackCollection) CreateNodeGroup(seq int, errs chan error) error {
+		return errors.Wrap(err, "error pasring JSON")
+	}
+
+	m := f.(map[string]interface{})
```
I once wrote https://godoc.org/github.com/errordeveloper/kubegen/pkg/macroproc#Tree, which we can extract, if seems appropriate... but there was also another package by someone else that I'm gonna try to find.

errordeveloper replied:
It could had been https://github.com/tidwall/gjson, and we would need it's setter counterpart https://github.com/tidwall/sjson. Let me know what you think! I'm happy to use my own as I know it well, and it'd be worse extorting it from kubegen, but it's entirely up to you =)

richardcase replied:
I'll take a look at them, thanks.

errordeveloper commented on pkg/cfn/manager/nodegroup.go:
```diff
@@ -28,6 +34,50 @@ func (c *StackCollection) CreateNodeGroup(seq int, errs chan error) error {
+	m := f.(map[string]interface{})
+	res := m["Resources"].(map[string]interface{})
+	ng := res["NodeGroup"].(map[string]interface{})
+	props := ng["Properties"].(map[string]interface{})
```
I'm not particularly picky, and we can probably assume schema won't change too much, but anyway there's an issue here that if we are to introduce check for each of the keys, it'd be a nightmare...

richardcase replied:
I think this would only be temporary until 'Refs' are handled completely in Goformation (or something else)

richardcase commented:
Closing this PR in favor of a more up to date one based on a cleaner branch
